**New Resource:** `azurerm_subscription` (#10718)
terraform-providers/terraform-provider-azurerm: website/docs/r/subscription.html.markdown

---
subcategory: "Base"
layout: "azurerm"
page_title: "Azure Resource Manager: azurerm_subscription"
description: |-
  Manages a Subscription.
---

# azurerm_subscription

Manages a Subscription.

> ~> **NOTE:** Destroying a Subscription controlled by this resource will place the Subscription into a cancelled state. It is possible to re-activate a subscription within 90-days of cancellation, after which time the Subscription is irrevocably deleted and the Subscription ID cannot be re-used. For further information see [here](https://docs.microsoft.com/en-us/azure/cost-management-billing/manage/cancel-azure-subscription#what-happens-after-subscription-cancellation)

Reviewer: also worth mentioning it's possible to actually "Delete" a subscription 3 days after it's cancelled, but TF can't do this for users

## Example Usage

```hcl
resource "azurerm_subscription" "example" {

>   alias              = "examplesub"

Reviewer: since this is a optional field, should we remove this:

```suggestion
```

~> **NOTE:** This value is only valid for Enterprise Agreements.

* `invoice_section` - (Optional) The Invoice Section name. Conflicts with `enrollment_account`, and `subscription_id`. Changing this forces a new Subscription to be created.

~> **NOTE:** This value is only valid for MCA and Partner Account types and must be used with `billing_profile`.

* `state` - (Optional) The target state of the subscription. Possible values are `Active` (default) and `Cancelled`.

* `subscription_id` - (Optional) The ID of the Subscription. Cannot be specified with `billing_account`, `billing_profile`, `enrollment_account`, or `invoice_section` Changing this forces a new Subscription to be created.

~> **NOTE:** This value can be specified only for adopting control of an existing Subscription, it cannot be used to provide a custom Subscription ID.

* `workload` - (Optional) The workload type of the Subscription.  Possible values are `Production` (default) and `DevTest`. Changing this forces a new Subscription to be created.

## Attributes Reference

In addition to the Arguments listed above - the following Attributes are exported: 

* `id` - The ID of the Subscription.

> * `tenant` - The ID of the Tenant to which the subscription belongs.

Reviewer: can we make this `tenant_id`

Author: It is in the resource, this is a typo 🙈

## Timeouts

The `timeouts` block allows you to specify [timeouts](https://www.terraform.io/docs/configuration/resources.html#timeouts) for certain actions:

* `create` - (Defaults to 30 minutes) Used when creating the Subscription.
* `read` - (Defaults to 5 minutes) Used when retrieving the Subscription.
* `update` - (Defaults to 30 minutes) Used when updating the Subscription.
* `delete` - (Defaults to 30 minutes) Used when deleting the Subscription.

## Import

Subscriptions can be imported using the `resource id`, e.g.

```shell
terraform import azurerm_subscription.example "/providers/Microsoft.Subscription/aliases/subscription1"
```

!> **NOTE:** When importing a Subscription that was not created programmatically, either by this Terraform resource or using the Alias API, it will have no Alias ID to import via `terraform import`.  

> In this scenario, the `subscription_id` property can be completed and Terraform will assume control of the existing subscription by creating an Alias. e.g.

Reviewer: presumably this wants to be:

```suggestion
In this scenario, the `subscription_id` property can be completed and Terraform will assume control of the existing subscription by updating the Subscription to add an Alias. e.g.
```

Author: It's a one-way reference. An Alias is a separate meta-resource that has a subscription ID reference, where the Subscription resource has no notion of Aliases, so I think this change might be misleading?

---
subcategory: "Base"
layout: "azurerm"
page_title: "Azure Resource Manager: azurerm_subscription"
description: |-
  Manages a Subscription.
---

# azurerm_subscription

> Manages a Subscription.

Reviewer: ```suggestion
Manages an Alias for a Subscription - which adds an Alias to an existing Subscription, allowing it to be managed in Terraform - or create a new Subscription with a new Alias.
```

~> **NOTE:** Destroying a Subscription controlled by this resource will place the Subscription into a cancelled state. It is possible to re-activate a subscription within 90-days of cancellation, after which time the Subscription is irrevocably deleted, and the Subscription ID cannot be re-used. For further information see [here](https://docs.microsoft.com/en-us/azure/cost-management-billing/manage/cancel-azure-subscription#what-happens-after-subscription-cancellation). Users can optionally delete a Subscription once 72 hours have passed, however, this functionality is not suitable for Terraform. A `Deleted` subscription cannot be reactivated.

> ~> **NOTE:** It is not possible to destroy (cancel) a subscription if it contains resources. If resources are present that are not managed by Terraform then these will need to be removed before the Subscription can be destroyed.

Reviewer: should we also add a note here about assuming a 1:1 between an alias and a subscription

* `create` - (Defaults to 30 minutes) Used when creating the Subscription.
* `read` - (Defaults to 5 minutes) Used when retrieving the Subscription.
* `update` - (Defaults to 30 minutes) Used when updating the Subscription.
* `delete` - (Defaults to 30 minutes) Used when deleting the Subscription.

## Import

Subscriptions can be imported using the `resource id`, e.g.

```shell
terraform import azurerm_subscription.example "/providers/Microsoft.Subscription/aliases/subscription1"
```

!> **NOTE:** When importing a Subscription that was not created programmatically, either by this Terraform resource or using the Alias API, it will have no Alias ID to import via `terraform import`.  
In this scenario, the `subscription_id` property can be completed and Terraform will assume control of the existing subscription by creating an Alias. e.g.

```hcl
// importing existing Manually Created Subscription with no Alias
resource "azurerm_subscription" "example" {
  alias             = "examplesub"
  subscription_name = "My Example Subscription"
  subscription_id   = "12345678-12234-5678-9012-123456789012"
}

> ```

Reviewer: maybe we should also move this into the example usage above - it's a single (small) resource so having multiple example usages should be fine?
